File: server/src/main/java/com/aymanetech/event/booking/application/service/BookingService.java
package com.aymanetech.event.booking.application.service;

import com.aymanetech.event.booking.application.dto.BookingRequestDto;
import com.aymanetech.event.booking.application.dto.BookingResponseDto;
import com.aymanetech.event.booking.domain.BookingId;
import com.aymanetech.event.booking.domain.BookingStatus;
import com.aymanetech.event.event.domain.vo.EventId;
import com.aymanetech.event.user.domain.vo.UserId;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface BookingService {
    BookingResponseDto createBooking(BookingRequestDto request);

    BookingResponseDto findBookingById(BookingId id);

    Page<BookingResponseDto> findBookingsByUser(UserId userId, Pageable pageable);

    Page<BookingResponseDto> findBookingsByEvent(EventId eventId, Pageable pageable);

    BookingResponseDto updateBookingStatus(BookingId bookingId, BookingStatus status);

    void cancelBooking(BookingId bookingId);

}

File: client/src/services/booking.service.ts
import apiClient from "../api/axios"
import type { BookingRequestDto, BookingResponseDto, BookingsPage, BookingStatus } from "../types/booking.types"

const API_PATH = "/v1/bookings"

export const BookingService = {
  createBooking: async (bookingData: BookingRequestDto): Promise<BookingResponseDto> => {
    const response = await apiClient.post<BookingResponseDto>(API_PATH, bookingData)
    return response.data
  },

  getBookingById: async (id: number): Promise<BookingResponseDto> => {
    const response = await apiClient.get<BookingResponseDto>(`${API_PATH}/${id}`)
    return response.data
  },

  getBookingsByUser: async (userId: number, pageNum: number = 0, pageSize: number = 10): Promise<BookingsPage> => {
    const response = await apiClient.get<BookingsPage>(
      `${API_PATH}/user/${userId}?pageNum=${pageNum}&pageSize=${pageSize}`
    )
    return response.data
  },

  getBookingsByEvent: async (eventId: number, pageNum: number = 0, pageSize: number = 10): Promise<BookingsPage> => {
    const response = await apiClient.get<BookingsPage>(
      `${API_PATH}/event/${eventId}?pageNum=${pageNum}&pageSize=${pageSize}`
    )
    return response.data
  },

  updateBookingStatus: async (id: number, status: BookingStatus): Promise<BookingResponseDto> => {
    const response = await apiClient.patch<BookingResponseDto>(
      `${API_PATH}/${id}/status?status=${status}`
    )
    return response.data
  },

  cancelBooking: async (id: number): Promise<void> => {
    await apiClient.delete(`${API_PATH}/${id}`)
  }
}

